### ISSUE TITLE
**`[FEATURE] Calculus Core Module - Derivatives, Integrals, Limits for Differentiable Programming`**

### ISSUE THREAD

**Issue by Gcarmnonapy7** [MorphosML/MorphosML#12]
# 📐 GitHub Issue: Calculus Core Module for MorphosML

---

## Issue Title:
**`[FEATURE] Calculus Core Module - Derivatives, Integrals, Limits for Differentiable Programming`**

---

### 🧮 Overview
This issue proposes the implementation of a comprehensive **Calculus Core Module** for MorphosML v0.3.0, providing foundational mathematical capabilities for differentiable programming, gradient-based optimization, and numerical methods. This will transform MorphosML from a classical ML library into a modern differentiable programming framework.

### 🎯 Goals
- Enable **automatic differentiation** for all existing models
- Provide **numerical and symbolic calculus** operations
- Support **gradient-based optimization** for neural networks
- Maintain **educational clarity** with step-by-step solvers
- Achieve **high performance** through expression templates and SIMD

---

#### 1.1 Automatic Differentiation Engine
- [ ] `Variable<T>` class with gradient tracking
- [ ] Computational graph node system
- [ ] Forward mode autodiff (for R→Rⁿ functions)
- [ ] Reverse mode autodiff (for Rⁿ→R functions - backpropagation)
- [ ] Gradient accumulation and zeroing
- [ ] `no_grad()` context manager
- [ ] `detach()` operation

**Files:** `calculus/autodiff.hpp`, `calculus/variable.hpp`, `calculus/graph.hpp`

#### 1.2 Basic Mathematical Operations with Derivatives
- [ ] Arithmetic: `+`, `-`, `*`, `/`, `-` (negation)
- [ ] Power operations: `pow()`, `sqrt()`, `square()`
- [ ] Exponential/Log: `exp()`, `log()`, `log10()`, `log1p()`
- [ ] Trigonometric: `sin()`, `cos()`, `tan()`, `asin()`, `acos()`, `atan()`
- [ ] Hyperbolic: `sinh()`, `cosh()`, `tanh()`, `asinh()`, `acosh()`, `atanh()`
- [ ] Comparison ops: `abs()`, `min()`, `max()`, `clamp()`

**Files:** `calculus/operations.hpp`, `calculus/elementary.hpp`

---

#### 2.1 Numerical Differentiation
- [ ] Finite difference methods (forward, backward, central, 5-point)
- [ ] Complex step differentiation
- [ ] Richardson extrapolation
- [ ] Higher-order derivatives (up to 10th order)
- [ ] Automatic step size selection

**Files:** `calculus/derivative/numerical.hpp`

#### 2.2 Symbolic Differentiation
- [ ] Derivative rule system (power, product, quotient, chain rules)
- [ ] Symbolic expression parser
- [ ] Derivative simplification
- [ ] Function composition derivatives

**Files:** `calculus/derivative/symbolic.hpp`

#### 2.3 Multivariate Calculus
- [ ] Gradient vector computation (∇f)
- [ ] Jacobian matrix (J_ij = ∂f_i/∂x_j)
- [ ] Hessian matrix (H_ij = ∂²f/∂x_i∂x_j)
- [ ] Directional derivatives (D_v f = ∇f · v)
- [ ] Laplacian operator (Δf = ∇·∇f)
- [ ] Partial derivatives (∂f/∂x_i)

**Files:** `calculus/derivative/multivariate.hpp`

#### 2.4 Activation Functions & Gradients
- [ ] Sigmoid & derivative: `σ(x)(1-σ(x))`
- [ ] Tanh & derivative: `sech²(x)`
- [ ] ReLU & derivative: `(x > 0 ? 1 : 0)`
- [ ] Leaky ReLU & derivative
- [ ] ELU & derivative
- [ ] Swish & derivative: `σ(x) + x·σ(x)(1-σ(x))`
- [ ] Softmax & Jacobian
- [ ] GELU & derivative (approximation)

**Files:** `calculus/activations.hpp`

---

#### 3.1 Numerical Integration Methods
- [ ] Newton-Cotes formulas:
  - [ ] Rectangular rules (left, right, midpoint)
  - [ ] Trapezoidal rule
  - [ ] Simpson's 1/3 rule
  - [ ] Simpson's 3/8 rule
  - [ ] Boole's rule
- [ ] Gaussian quadrature:
  - [ ] Gauss-Legendre (finite intervals)
  - [ ] Gauss-Laguerre ([0, ∞))
  - [ ] Gauss-Hermite ((-∞, ∞))
- [ ] Adaptive quadrature with error control
- [ ] Romberg integration (Richardson extrapolation)

**Files:** `calculus/integral/numerical.hpp`, `calculus/integral/quadrature.hpp`

#### 3.2 Advanced Integration
- [ ] Double integrals (∬)
- [ ] Triple integrals (∭)
- [ ] Line integrals (∫_C f ds)
- [ ] Surface integrals (∬_S f dS)
- [ ] Improper integrals (Type I & II)
- [ ] Cauchy principal value
- [ ] Monte Carlo integration (high dimensions)
- [ ] Importance sampling
- [ ] Variance reduction techniques

**Files:** `calculus/integral/advanced.hpp`, `calculus/integral/monte_carlo.hpp`

#### 3.3 Integration Utilities
- [ ] Integration of tabular data (discrete points)
- [ ] Error estimation
- [ ] Convergence acceleration (Wynn's epsilon)
- [ ] Contour integration (complex analysis)

**Files:** `calculus/integral/utils.hpp`

---

#### 4.1 Limit Computation
- [ ] Two-sided limits: `lim_{x→a} f(x)`
- [ ] One-sided limits: `lim_{x→a⁻} f(x)`, `lim_{x→a⁺} f(x)`
- [ ] Limits at infinity: `lim_{x→∞} f(x)`, `lim_{x→-∞} f(x)`
- [ ] Sequence limits: `lim_{n→∞} a_n`

**Files:** `calculus/limit/basic.hpp`

#### 4.2 Limit Techniques
- [ ] Direct substitution
- [ ] Factoring (0/0 forms)
- [ ] Rationalization
- [ ] L'Hôpital's rule (0/0, ∞/∞)
- [ ] Series expansion
- [ ] Squeeze theorem implementation

**Files:** `calculus/limit/techniques.hpp`

#### 4.3 Advanced Limit Topics
- [ ] Indeterminate form detection (7 types)
- [ ] Asymptotic analysis (Big-O, small-o)
- [ ] Multivariate limits (2D, 3D)
- [ ] Limit along a path
- [ ] Path independence checking
- [ ] Iterated limits

**Files:** `calculus/limit/advanced.hpp`, `calculus/limit/multivariate.hpp`

#### 4.4 Applications
- [ ] Continuity testing
- [ ] Discontinuity classification
- [ ] Differentiability checking
- [ ] Radius of convergence for power series
- [ ] Series convergence tests

**Files:** `calculus/limit/applications.hpp`

---

#### 5.1 Differentiable Loss Functions
- [ ] Mean Squared Error (MSE) + gradient
- [ ] Mean Absolute Error (MAE) + subgradient
- [ ] Binary Cross Entropy + gradient
- [ ] Categorical Cross Entropy + gradient
- [ ] Huber Loss + gradient
- [ ] Hinge Loss + subgradient
- [ ] KL Divergence + gradient
- [ ] Cosine Similarity Loss

**Files:** `calculus/loss.hpp`

#### 5.2 Gradient-Based Optimizers
- [ ] Stochastic Gradient Descent (SGD)
- [ ] SGD with Momentum
- [ ] Nesterov Accelerated Gradient (NAG)
- [ ] AdaGrad
- [ ] RMSProp
- [ ] Adam (with bias correction)
- [ ] AdamW (weight decay)
- [ ] Nadam
- [ ] AMSGrad
- [ ] Learning rate schedulers (step, cosine, exponential)

**Files:** `calculus/optimizers.hpp`, `calculus/schedulers.hpp`

---

#### 6.1 Unit Tests
- [ ] Autodiff correctness (vs analytical derivatives)
- [ ] Numerical vs analytical gradient comparison
- [ ] Loss function gradients
- [ ] Optimizer convergence tests
- [ ] Edge cases (NaN, Inf, underflow/overflow)
- [ ] Memory leak detection for computational graph

**Files:** `tests/calculus/test_autodiff.cpp`, `tests/calculus/test_optimizers.cpp`

#### 6.2 Validation Suite
- [ ] Compare with PyTorch/TensorFlow results
- [ ] Gradient checking for random functions
- [ ] Optimization convergence benchmarks
- [ ] Numerical stability testing
- [ ] Performance profiling

**Files:** `tests/calculus/validation.cpp`, `benchmarks/calculus/benchmark.cpp`

#### 6.3 Integration Tests
- [ ] Neural network training end-to-end
- [ ] Integration with existing BaseModel
- [ ] Multi-threaded gradient computation
- [ ] Serialization of computational graph

**Files:** `tests/calculus/integration.cpp`

---

#### 7.1 Step-by-Step Solver
- [ ] Limit solver with detailed steps
- [ ] Derivative solver with rule explanations
- [ ] Integral solver with method selection
- [ ] Solution visualization

**Files:** `calculus/education/solver.hpp`

#### 7.2 Interactive Tutorials
- [ ] Calculus for ML tutorial
- [ ] Automatic differentiation explained
- [ ] Gradient descent visualization
- [ ] Common pitfalls and debugging guide

**Files:** `docs/calculus/tutorials/`

#### 7.3 Examples
- [ ] `examples/neural_network_from_scratch.cpp`
- [ ] `examples/automatic_differentiation_demo.cpp`
- [ ] `examples/optimizer_comparison.cpp`
- [ ] `examples/calculus_visualization.cpp`

---

### Core Class Structure
```cpp
namespace morphosml::calculus {

template<typename T>
class Variable {
    T data_;
    std::shared_ptr<Variable<T>> grad_;
    std::vector<std::shared_ptr<Variable<T>>> parents_;
    std::function<void()> backward_fn_;
    bool requires_grad_;
    std::string name_;  // for debugging
    
    // Operations
    Variable<T> operator+(const Variable<T>& other);
    Variable<T> operator*(const Variable<T>& other);
    // ... other operators
};

class Graph {
    std::vector<Variable*> topological_order();
    void backward(Variable& loss);
    void zero_grad();
};

template<typename T>
class Optimizer {
    virtual void step() = 0;
    virtual void zero_grad();
};

// Loss function base
template<typename T>
class LossFunction {
    virtual Variable<T> compute(const Variable<T>& pred, 
                                const Variable<T>& target) = 0;
};
}
```

### Integration with Existing BaseModel
```cpp
template<typename T>
class DifferentiableModel : public BaseModel<T> {
protected:
    std::vector<Variable<T>> parameters_;
    Optimizer<T>* optimizer_;
    LossFunction<T>* loss_fn_;
    
public:
    void fit(const std::vector<std::vector<T>>& X,
             const std::vector<T>& y) override {
        // Convert to Variables
        auto X_var = to_variable(X);
        auto y_var = to_variable(y);
        
        // Forward pass with autodiff
        auto output = forward(X_var);
        auto loss = loss_fn_->compute(output, y_var);
        
        // Backward pass
        loss.backward();
        
        // Optimize
        optimizer_->step();
        optimizer_->zero_grad();
    }
};
```

---

## 📊 Performance Targets

| Operation | Target Speed | Target Accuracy |
|-----------|--------------|-----------------|
| 1st derivative (autodiff) | < 1μs | 1e-14 |
| 1st derivative (numerical) | < 0.5μs | 1e-8 |
| Jacobian (10x10) | < 10μs | 1e-12 |
| Hessian (10x10) | < 50μs | 1e-10 |
| Definite integral (adaptive) | < 100μs | 1e-9 |
| Double integral (MC, 1e5 samples) | < 100ms | 1e-3 |
| Limit computation | < 10μs | 1e-10 |
| Optimizer step (Adam, 1e6 params) | < 1ms | - |

---

### Technical Specifications
- **Language**: C++17 (or newer)
- **Dependencies**: 
  - Standard library only (no external deps for core)
  - Optional: Eigen for advanced linear algebra
- **Memory**: RAII, smart pointers for graph nodes
- **Thread Safety**: Document assumptions
- **Error Handling**: Exceptions for math errors (division by zero, etc.)

### Code Quality
- [ ] 90%+ test coverage
- [ ] Clang-format with custom style
- [ ] Clang-tidy static analysis passing
- [ ] AddressSanitizer and UndefinedBehaviorSanitizer clean
- [ ] Documentation for all public APIs

---

## 📚 Documentation Requirements

- [ ] API reference (Doxygen)
- [ ] Tutorial: "Calculus for Machine Learning"
- [ ] Tutorial: "Automatic Differentiation Explained"
- [ ] Tutorial: "Training Neural Networks from Scratch"
- [ ] Examples directory with 5+ complete examples
- [ ] Performance optimization guide
- [ ] Common pitfalls and debugging strategies
- [ ] Integration guide with existing MorphosML models

---

### Unit Tests (Required for all functions)
```cpp
TEST(AutodiffTest, BasicArithmetic) {
    auto x = Variable<double>(2.0, requires_grad=true);
    auto y = x * x + 3 * x + 2;
    y.backward();
    
    EXPECT_NEAR(x.grad(), 7.0, 1e-12);  // 2x + 3 at x=2
}
```

### Validation Tests
- Compare with analytical derivatives for 100 random functions
- Verify against PyTorch results for same computation
- Optimization convergence on standard test functions (Rosenbrock, etc.)

### Benchmark Suite
- Microbenchmarks for each operation
- Macrobenchmarks for end-to-end training
- Memory usage profiling

---

## 🚀 Stretch Goals (Post v0.3.0)

- [ ] GPU acceleration (CUDA/HIP)
- [ ] Sparse automatic differentiation
- [ ] Higher-order gradients (grad², grad³)
- [ ] Neural ODE solvers
- [ ] Differentiable physics engine
- [ ] Automatic sparse Jacobian detection
- [ ] Checkpointing for long computational chains
- [ ] Distributed autodiff (MPI)

---

## ✅ Acceptance Criteria

1. All planned features implemented and tested
2. 90%+ test coverage across calculus module
3. No performance regressions in existing models
4. All examples compile and run successfully
5. Documentation complete with tutorials
6. Integration with existing BaseModel works seamlessly
7. Can train a neural network on MNIST with 95%+ accuracy
8. Benchmark suite shows expected performance

---

## 👥 Contributors Needed

- **Core Maintainer** (lead developer)
- **Autodiff Specialist** (graph algorithms)
- **Numerical Methods Expert** (integration/limits)
- **Documentation Writer** (tutorials)
- **Test Engineer** (validation suite)

---

## 📅 Timeline

- **Week 1-2**: Phase 1 (Core Infrastructure)
- **Week 3-4**: Phase 2 (Derivative Calculus)
- **Week 5-6**: Phase 3 (Integral Calculus)
- **Week 7**: Phase 4 (Limit Calculus)
- **Week 8**: Phase 5 (Loss & Optimizers)
- **Week 9**: Phase 6 (Testing & Validation)
- **Week 10**: Phase 7 (Educational Features)
- **Week 11**: Documentation polish & release preparation

**Target Release**: v0.3.0 (approx. 3 months from start)

---

## 🔗 Related Issues

- #42: Performance Optimization Roadmap
- #15: Neural Network Module
- #23: Automatic Differentiation Discussion
- #67: Educational Tools Enhancement

---

## 💬 Discussion Points

1. Should we support complex numbers for autodiff?
2. What level of symbolic simplification should we implement?
3. Should we integrate with Eigen for matrix operations?
4. How to handle graph optimization (common subexpression elimination)?
5. What's the priority: numerical stability or performance?

---

## 📝 Labels

`enhancement`, `calculus-core`, `differentiable-programming`, `v0.3.0`, `high-priority`

---

**/assign @morphosml-maintainer**
**/priority high**
**/size x-large**
